File: packages/elixir/lib/kreuzberg/code_comment_info.ex
defmodule Kreuzberg.CodeCommentInfo do
  @moduledoc """
  Comment information.

  ## Fields

    * `:text` - Comment text
    * `:kind` - Comment kind (e.g. "line", "block")
    * `:span` - Source span
  """

  @type t :: %__MODULE__{
          text: String.t(),
          kind: String.t(),
          span: Kreuzberg.CodeSpan.t()
        }

  defstruct text: "",
            kind: "",
            span: %Kreuzberg.CodeSpan{}

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    span =
      case Map.get(data, "span", Map.get(data, :span)) do
        %Kreuzberg.CodeSpan{} = s -> s
        map when is_map(map) -> Kreuzberg.CodeSpan.from_map(map)
        _ -> %Kreuzberg.CodeSpan{}
      end

    %__MODULE__{
      text: Map.get(data, "text", Map.get(data, :text, "")),
      kind: Map.get(data, "kind", Map.get(data, :kind, "")),
      span: span
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = info) do
    %{
      "text" => info.text,
      "kind" => info.kind,
      "span" => Kreuzberg.CodeSpan.to_map(info.span)
    }
  end
end

File: kreuzberg/packages/php/src/Types/CodeCommentInfo.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Comment information.
 *
 * @property-read string $text Comment text
 * @property-read string $kind Comment kind (e.g. 'line', 'block')
 * @property-read CodeSpan $span Source span
 */
readonly class CodeCommentInfo
{
    public function __construct(
        public string $text,
        public string $kind,
        public CodeSpan $span,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $text */
        $text = $data['text'] ?? '';

        /** @var string $kind */
        $kind = $data['kind'] ?? '';

        /** @var array<string, mixed> $spanData */
        $spanData = $data['span'] ?? [];

        return new self(
            text: $text,
            kind: $kind,
            span: CodeSpan::fromArray($spanData),
        );
    }
}
